include/onnxruntime/core/framework/custom_execution_provider.h

```diff
@@ -0,0 +1,354 @@
+  ParseArgs() {
+    return TypeConstraint("T", DataType::float_tp).ParseArgs<ith_input + 1, ith_output, Ts...>();
+  }
+
+  template <size_t ith_input, size_t ith_output, typename T, typename... Ts>
+  typename std::enable_if<std::is_same<T, const TensorT1<float>&>::value, IKernelBuilder&>::type
+  ParseArgs() {
+    return TypeConstraint("T1", DataType::float_tp).ParseArgs<ith_input + 1, ith_output, Ts...>();
+  }
+
+  template <size_t ith_input, size_t ith_output, typename T, typename... Ts>
+  typename std::enable_if<std::is_same<T, const TensorV<float>&>::value, IKernelBuilder&>::type
+  ParseArgs() {
+    return TypeConstraint("V", DataType::float_tp).ParseArgs<ith_input + 1, ith_output, Ts...>();
+  }
+
+  template <size_t ith_input, size_t ith_output, typename T, typename... Ts>
+  typename std::enable_if<std::is_same<T, Aliased<float>&>::value, IKernelBuilder&>::type
+  ParseArgs() {
+    return Alias(Aliased<float>::InputIndice(), 0).ParseArgs<ith_input, ith_output + 1, Ts...>();
+  }
+
+  template <typename... Args>
+  IKernelBuilder& ParseFn(onnxruntime::Status (*compute_fn)(Args...)) {
+    using KernelType = FnKernel<Args...>;
+    auto fn = [](IKernelInfo& info) -> IKernel {
+      std::make_unique<KernelType>(info, compute_fn);
+        }
+    //kernel_ = std::make_unique<KernelType>(compute_fn);
+    return ParseArgs<0, 0, Args...>();
+    //KernelBulder<KernelType> builder;
+    //return ParseArgs<0, 0, Args...>(builder);
+  }
+
+  template<typename K, typename... Args>
+  IKernelBuilder& ParseStruct(onnxruntime::Status (K::*compute_fn)(Args...)) {
+    using KernelType = StructKernel<K>;
+    kernel_ = std::make_unique<KernelType>();
+    return ParseArgs<0, 0, Args...>();
+  }
+
+  std::unique_ptr<IKernel> kernel_;
+};
+
+struct IKernelRegistry {
+  virtual ~IKernelRegistry() = default;
+  virtual IKernelBuilder& CreateBuilder() = 0;
+  template <typename... Args>
+  IKernelBuilder& RegisterKernel(const char* ep,
+                                 const char* domain,
+                                 const char* op,
+                                 int since_ver,
+                                 int end_ver,
+                                 onnxruntime::Status (*compute_fn)(Args...)) {
+    return CreateBuilder().Provider(ep).SetDomain(domain).SetName(op).SinceVersion(since_ver, end_ver).ParseFn<Args...>(compute_fn);
+  }
+  template <typename K, typename... Args>
+  IKernelBuilder& RegisterKernel(const char* ep,
+                                 const char* domain,
+                                 const char* op,
+                                 int since_ver,
+                                 int end_ver) {
+    return CreateBuilder().Provider(ep).SetDomain(domain).SetName(op).SinceVersion(since_ver, end_ver).ParseStruct<K, Args...>(&K::Compute);
+  }
+};
+
+/////////////////////////////////////////////////// Ep Interface ///////////////////////////////////////////////////
+
+struct IExecutionProvider {
```

Comment:
I think there is some dup code here, especially this class, which I have moved to include/onnxruntime/interface/provider/provider.h and renamed as ExecutionProvider